### ISSUE TITLE
Invalid xpcall fails silently when neovim is built against lua 5.1

### ISSUE THREAD

**Issue by subnut** [neovim/neovim#27289]
### TL;DR
`xpcall` doesn't fail on LuaJIT when called as
```lua
xpcall(func, handler, arg1, arg2, ...)
```
but _does_ fail when compiled agains Lua 5.1

---

Assume the following lua code.
```lua
function add(a, b)
  return a + b
end
function handler(err)
  print(err)
end
```

Now, `:h luaref-xpcall()` states -
```vimhelp
xpcall({f}, {err})                                             *luaref-xpcall()*
        This function is similar to `pcall` (see |luaref-pcall()|), except that
        you can set a new error handler.
```
`:h luaref-pcall()`
```vimhelp
pcall({f}, {arg1}, {...})                                       *luaref-pcall()*
        Calls function {f} with the given arguments in `protected mode`. This
```

So according to the documentation, this is valid lua code -
```lua
pcall(add, 1, 2)
```

So far so good.

---
Now here comes the problem. 
 
Note that the `xpcall` documentation said
> This function is similar to `pcall`, except that
>        you can set a new error handler.

So people may assume that it means xpcall is the same as pcall, except with an error handler function after `func` and **before the arguments.**

Now, the function signature of xpcall is written as `xpcall(f, err)`, (which is correct)  
but this piece of information directly contradicts the previous assumption made by the reader!

So what do they do? They, like a sensible person, test it out for themselves!  
They download an official version of neovim (from any of the relevant sources) and run -
```lua
print(xpcall(add, handler, 1, 2))
```
it prints -
```
true
3
```
So, they conclude their assumption was correct! `xpcall` _does_ take extra arguments (that are then passed to `f`). The documention just hasn't been updated, they think. And they go there merry way making lua plugins.

Now what they _didn't_ know is that the same `xpcall` call would **fail** if neovim was compiled with the official Lua 5.1 release. They write a plugin based on that false assumption, and somebody writes their own plugin that depends on that plugin,  
...  
and now suddenly there's a neovim plugin that fails _silently_ when it's run on a neovim distribution that's compiled on Lua 5.1 (probably due to LuaJIT not being available).

The plugin user files a bug report, but neither parties have any idea what might be failing, because there's no error message! How am I supposed to debug an error if there _is_ no error?!

### Steps to reproduce

See above.

### Expected behavior

Two options -
1. Make LuaJIT-compiled neovim's `xpcall` fail when provided more/less than 2 arguments
2. Make Lua-compiled neovim transparently accept the additional arguments, and update documention to reflect the same.

**NOTE:**  
Just updating the documention without fixing the behaviour should NOT be an option IMHO/

### Neovim version (nvim -v)

0.9.5

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

any

### Terminal name/version

any

### $TERM environment variable

any

### Installation

any

**Comment by clason** [comment]
I don't see what this has to do with Neovim; this is purely about LuaJIT/PUC Lua differences. You get the interpreter you (or the person who built for you) chose.

**Comment by clason** [comment]
****

**Comment by lewis6991** [comment]
As long as there isn't any code in our runtime that depends on this, there isn't an issue here.
